docs/community/tutorials/bert-qa-benchmarking/index-bert-qa-benchmarking-with-redisai-and-redisgears.mdx

```diff
@@ -0,0 +1,406 @@
+
+```python
+    token_key = f"tokenized:bert:qa:{sentence_key}"
+    # encode question
+    input_ids_question = tokenizer.encode(question, add_special_tokens=True, truncation=True, return_tensors="np")
+    t=redisAI.getTensorFromKey(token_key)
+    input_ids_context=to_np(t,np.int64)
+    # merge (append) with potential answer, context - is pre-tokenized paragraph
+    input_ids = np.append(input_ids_question,input_ids_context)
+    attention_mask = np.array([[1]*len(input_ids)])
+    input_idss=np.array([input_ids])
+    num_seg_a=input_ids_question.shape[1]
+    num_seg_b=input_ids_context.shape[0]
+    token_type_ids = np.array([0]*num_seg_a + [1]*num_seg_b)
+    # create actual model runnger for RedisAI
+    modelRunner = redisAI.createModelRunner(f'bert-qa{hash_tag}')
+    # make sure all types are correct
+    input_idss_ts=redisAI.createTensorFromBlob('INT64', input_idss.shape, input_idss.tobytes())
+    attention_mask_ts=redisAI.createTensorFromBlob('INT64', attention_mask.shape, attention_mask.tobytes())
+    token_type_ids_ts=redisAI.createTensorFromBlob('INT64', token_type_ids.shape, token_type_ids.tobytes())
+    redisAI.modelRunnerAddInput(modelRunner, 'input_ids', input_idss_ts)
+    redisAI.modelRunnerAddInput(modelRunner, 'attention_mask', attention_mask_ts)
+    redisAI.modelRunnerAddInput(modelRunner, 'token_type_ids', token_type_ids_ts)
+    redisAI.modelRunnerAddOutput(modelRunner, 'answer_start_scores')
+    redisAI.modelRunnerAddOutput(modelRunner, 'answer_end_scores')
+    # run RedisAI model runner
+    res = await redisAI.modelRunnerRunAsync(modelRunner)
+    answer_start_scores=to_np(res[0],np.float32)
+    answer_end_scores = to_np(res[1],np.float32)
+    answer_start = np.argmax(answer_start_scores)
+    answer_end = np.argmax(answer_end_scores) + 1
+    answer = tokenizer.convert_tokens_to_string(tokenizer.convert_ids_to_tokens(input_ids[answer_start:answer_end],skip_special_tokens = True))
+    log("Answer "+str(answer))
+    return answer
+
+```
+
+Full [code in github](https://github.com/applied-knowledge-systems/the-pattern-api/blob/156633b9934f1243775671ce6c18ff2bf471c0ce/qasearch/qa_redisai_keymiss_no_cache_np.py#L34)
+
+Overall call for BERT QA API looks like this:
+
+![Architecture Diagram for BERT QA RedisGears and RedisAI](diagrams_02.png)
+
+Where I use two RedisGears cool features: capture event on key miss and async/await with the ability to run RedisAI on each shard and in non-locking for primary thread way - so Redis Cluster can continue to serve other customers. For benchmarks, caching RedisAI response is [disabled](https://github.com/applied-knowledge-systems/the-pattern-api/blob/156633b9934f1243775671ce6c18ff2bf471c0ce/qasearch/qa_redisai_keymiss_no_cache_np.py#L29) and if you are getting response time in nanoseconds on the second call rather then milliseconds you have deployed production version of RedisGears.
```

Comment:
```suggestion
Here I used two cool features of RedisGears: capture event on key miss and async/await with the ability to run RedisAI on each shard and in a non-locking way - so that Redis Cluster can continue to serve other clients. For benchmarks, caching RedisAI responses is [disabled](https://github.com/applied-knowledge-systems/the-pattern-api/blob/156633b9934f1243775671ce6c18ff2bf471c0ce/qasearch/qa_redisai_keymiss_no_cache_np.py#L29) and if you are getting response time in nanoseconds on the second call rather then milliseconds you have deployed production version of RedisGears.
```

The last part of this sentence needs work - not sure what it means.